### PR TITLE
fallback to Buffer.isUtf8 on platforms without icu

### PR DESCRIPTION
**Pull request by KhafraDev**
## Status

- [ ] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [ ] Tested
- [ ] Benchmarked (**optional**)
- [ ] Documented
- [ ] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: lib/web/websocket/util.js [hunk 1]
```diff
@@ -200,6 +201,33 @@ function failWebsocketConnection (ws, reason) {
...
+ * @param {Buffer} buffer
+ */
+function utf8Decode (buffer) {
+  if (hasIntl) {
+    return fatalDecoder.decode(buffer)
+  } else {
+    if (!isUtf8?.(buffer)) {
+      // TODO: remove once node 18 or < node v18.14.0 is dropped
+      if (!isUtf8) {
+        process.emitWarning('ICU is not supported and no fallback exists. Please upgrade to at least Node v18.14.0.', {
```
Comment (metcoder95) [inline lib/web/websocket/util.js:219]: Shall we reduce the emits to just `once`?

**Review by mcollina — CHANGES_REQUESTED** [review]
we need tests

**Comment by KhafraDev** [comment]
https://github.com/nodejs/undici/issues/3002#issuecomment-2022910249

In the meantime, if this doesn't work I don't think there's any harm. WebSocket doesn't currently work for those using node without ICU. There isn't a case where it'll allow invalid utf-8: if someone's node version is without icu and doesn't support buffer.isUtf8, an error is thrown.

**Comment by KhafraDev** [comment]
@mcollina I confirmed locally that this PR works on a --without-intl build. Tbh I think I'll end up upstreaming this change so TextDecoder in the non-icu build can also use the fatal option.

**Comment by mweberxyz** [comment]
FWIW- I used this branch as a test for my in-progress automated `--without-intl` CI branch and undici tests (both in main and this branch) are still failing.

See: https://github.com/mweberxyz/nodejs-undici/actions/runs/8474734954/job/23221671648?pr=5

**Comment by mweberxyz** [comment]
FWIW - also passes my proposed automated w/o intl ci checks: https://github.com/mweberxyz/nodejs-undici/pull/7

File: lib/web/websocket/util.js [hunk 2]
```diff
@@ -200,6 +201,33 @@ function failWebsocketConnection (ws, reason) {
...
 
+// https://nodejs.org/api/intl.html#detecting-internationalization-support
+const hasIntl = typeof process.versions.icu === 'string'
+const fatalDecoder = hasIntl ? new TextDecoder('utf-8', { fatal: true }) : undefined
+
+/**
+ * Converts a Buffer to utf-8, even on platforms without icu.
+ * @param {Buffer} buffer
+ */
+function utf8Decode (buffer) {
```
Comment (Uzlopak) [inline lib/web/websocket/util.js:212]: const utf8Decode = hasIntl ? fatalDecode.decode
: etc.

**Review by mcollina — APPROVED** [review]
lgtm with others comments fixed

**Pull request merged**